pkg/canary/writer/push.go

```diff
@@ -0,0 +1,208 @@
+		}
+		client.Transport = rt
+		scheme = "https"
+	}
+
+	u := url.URL{
+		Scheme: scheme,
+		Host:   lokiAddr,
+		Path:   pushEndpoint,
+	}
+
+	return &Push{
+		lokiURL:     u.String(),
+		tenantID:    tenantID,
+		httpClient:  client,
+		userAgent:   defaultUserAgent,
+		contentType: defaultContentType,
+		logger:      logger,
+		labelName:   labelName,
+		labelValue:  labelValue,
+		streamName:  streamName,
+		streamValue: streamValue,
+		username:    username,
+		password:    password,
+	}, nil
+}
+
+// Write implements the io.Writer.
+func (p *Push) Write(payload []byte) (int, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), p.httpClient.Timeout)
+	defer cancel()
+	if err := p.send(ctx, payload); err != nil {
+		return 0, err
+	}
+	return len(payload), nil
+}
+
+func (p *Push) parsePayload(payload []byte) (*logproto.PushRequest, error) {
+	// payload that is sent by the `writer` will be in format `LogEntry`
+	var (
+		tsStr, logLine string
+	)
+	if _, err := fmt.Sscanf(string(payload), LogEntry, &tsStr, &logLine); err != nil {
+		return nil, fmt.Errorf("failed to parse payload written sent by writer: %w", err)
+	}
+
+	ts, err := strconv.ParseInt(tsStr, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse unix nano timestamp: %w", err)
+	}
+
+	labels := model.LabelSet{
+		model.LabelName(p.labelName):  model.LabelValue(p.labelValue),
+		model.LabelName(p.streamName): model.LabelValue(p.streamValue),
+	}
+
+	return &logproto.PushRequest{
+		Streams: []logproto.Stream{
+			{
+				Labels: labels.String(),
+				Entries: []logproto.Entry{
+					{
+						Timestamp: time.Unix(0, ts),
+						Line:      string(payload),
+					},
+				},
+				Hash: uint64(labels.Fingerprint()),
+			},
+		},
+	}, nil
+}
+
+// send does the heavy lifting of sending the generated logs into the Loki server.
+// It won't batch.
+func (p *Push) send(ctx context.Context, payload []byte) error {
+	preq, err := p.parsePayload(payload)
+	if err != nil {
+		return err
+	}
+
+	payload, err = proto.Marshal(preq)
+	if err != nil {
+		return fmt.Errorf("failed to marshal payload to json: %w", err)
+	}
+
+	payload = snappy.Encode(nil, payload)
+
+	req, err := http.NewRequest("POST", p.lokiURL, bytes.NewReader(payload))
+	if err != nil {
+		return fmt.Errorf("failed to create push request: %w", err)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", p.contentType)
+	req.Header.Set("User-Agent", p.userAgent)
+
+	// set org-id
+	if p.tenantID != "" {
+		req.Header.Set("X-Scope-OrgID", p.tenantID)
+	}
+
+	// basic auth if provided
+	if p.username != "" {
+		req.SetBasicAuth(p.username, p.password)
+	}
+
+	resp, err := p.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to push payload: %w", err)
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			level.Error(p.logger).Log("msg", "failed to close response body", "error", err)
+		}
+	}()
+
+	if resp.StatusCode/100 != 2 {
```

Comment:
We should make it very explicit that this push mode does not retry when rate-limited, or when receiving 5xx responses.

I'm skeptical about this, though, because it means that the canary may show that a Loki cell is unhealthy even when things are fine, because of a transient error or a rate-limit.

Reply:
Yea right. there is no backoff or retry mechanism in this push mode currently. that was intentional for the same purpose you mentioned. It's just a normal http client with timeout. curious what made you to suspect this would retry?

or do you want to add some comment for clarification?

Reply:
>I guess because any transient failure in the Loki cell which self-heals will look like a failure to the canary, and give the operator an incorrect impression

So I assume you want to have retry mechanism similar to promtail. sounds reasonable. will add it

Reply:
Yeah maybe an exponential backoff with a configurable number of retries
Doesn't have to be too fancy

Reply:
Done :+1: